cmd/traffic/cmd/manager/state/workload_info_watcher.go

```diff
@@ -171,7 +174,60 @@ func rpcWorkloadState(s workload.State) (state rpc.WorkloadInfo_State) {
 	return state
 }
 
-func rpcWorkload(wl k8sapi.Workload, as rpc.WorkloadInfo_AgentState, iClients []*rpc.WorkloadInfo_Intercept) *rpc.WorkloadInfo {
+func extractPortsFromSidecar(_ context.Context, sc *agentconfig.Sidecar) []*rpc.WorkloadPortInfo {
+	if sc == nil {
+		return nil
+	}
+
+	var ports []*rpc.WorkloadPortInfo
+	seenPorts := make(map[types.PortAndProto]bool) // Track seen port+protocol combinations to avoid duplicates
+
+	// Extract ports from all containers in the sidecar
+	for _, container := range sc.Containers {
+		for _, intercept := range container.Intercepts {
+			// Create a unique key using port number and protocol
+			// This allows the same port with different protocols (TCP/UDP)
+			portKey := types.PortAndProto{
+				Port:  intercept.ServicePort,
+				Proto: intercept.Protocol,
+			}
```

Comment:
This will only work when there actually is a service port. Headless workloads won't have this. I suggest that you use the container port and proto as key instead. It's guaranteed to be present.